### PR TITLE
[database] Set storage load limit from the channel

### PR DESCRIPTION
**Pull request by clpolet**
* Set load limit when loading from the storage
* Use SQL limit to limit number of logs loaded from the DB
* Some renaming, wording
* Introduce MSDBStoragePrivate.h to expose members to tests
* Update unit tests
* Add unit tests to cover logs loading from DB with limit

### CODE REVIEW

File: MobileCenter/MSDBStorageTests.m [hunk 1]
```diff
@@ -1,12 +1,16 @@
+#import "MSAbstractLog.h"
+#import "MSDBStoragePrivate.h"
+#import "MSDatabaseConnection.h"
```
Comment (TroubleMakerBen) [inline MobileCenter/MSDBStorageTests.m]: The order for the imports is 
`#import <...>`
`#import "..."`

we stopped being consistent, let's do it again!

File: MobileCenter/MobileCenter/Internals/Channel/MSChannelDefault.m [hunk 2]
```diff
@@ -347,6 +348,7 @@ - (void)deleteAllLogsWithErrorSync:(NSError *)error {
     [self.storage deleteLogsForId:batchId withGroupID:self.configuration.groupID];
   }
 
+  // TODO: What about the database schema, it will stay on disc even after disabling?
```
Comment (TroubleMakerBen) [inline MobileCenter/MobileCenter/Internals/Channel/MSChannelDefault.m]: I'd say if the schema stays on disk and the DB is otherwise empty we're good.

File: MobileCenter/MobileCenter/Internals/Model/Utils/MSUserDefaults.m [hunk 3]
```diff
@@ -30,36 +30,37 @@ - (void)removeObjectForKey:(NSString *)key {
...
   NSMutableDictionary *update = [[NSMutableDictionary alloc] initWithDictionary:dict];
 
-  /* Get from local store */
+  /* Get from local store. */
   NSDictionary *store = [[NSUserDefaults standardUserDefaults] dictionaryForKey:key];
   CFAbsoluteTime ts = [store[kMSUserDefaultsTs] doubleValue];
   MSLogVerbose([MSMobileCenter logTag], @"Settings:store[%@]=%@", key, store);
 
-  /* Force update if timestamp expiration is reached */
+  /* Force update if timestamp expiration is reached. */
```
Comment (TroubleMakerBen) [inline MobileCenter/MobileCenter/Internals/Model/Utils/MSUserDefaults.m]: Use `//` here =)

File: MobileCenter/MobileCenter/Internals/Model/Utils/MSUserDefaults.m [hunk 4]
```diff
@@ -30,36 +30,37 @@ - (void)removeObjectForKey:(NSString *)key {
...
   if (ts <= 0.0 || expiration <= 0.0f || fabs(CFAbsoluteTimeGetCurrent() - ts) < (double)expiration) {
-    /* Remove if already in store and value is the same */
+    
+    /* Remove if already in store and value is the same. */
```
Comment (TroubleMakerBen) [inline MobileCenter/MobileCenter/Internals/Model/Utils/MSUserDefaults.m]: Use `//` here

File: MobileCenter/MobileCenter/Internals/Model/Utils/MSUserDefaults.m [hunk 5]
```diff
@@ -30,36 +30,37 @@ - (void)removeObjectForKey:(NSString *)key {
...
     for (NSString *k in [store allKeys]) {
       if (update[k] != nil && [update[k] isEqual:store[k]])
         [update removeObjectForKey:k];
     }
   }
 
-  /* If still values to update */
+  /* If still values to update. */
```
Comment (TroubleMakerBen) [inline MobileCenter/MobileCenter/Internals/Model/Utils/MSUserDefaults.m]: Use `//` here (I'll stop adding the same comment now ;))

File: MobileCenter/MobileCenter/Internals/Storage/MSDBStorage.m [hunk 6]
```diff
@@ -45,56 +38,80 @@ - (BOOL)saveLog:(id<MSLog>)log withGroupID:(NSString *)groupID {
...
+  NSString *addLogQuery = [NSString
+      stringWithFormat:@"insert or replace into %@ values ('%@', '%@')", kMSLogTableName, groupID, base64Data];
   return [self.connection executeQuery:addLogQuery];
 }
 
 - (NSArray<MSLog> *)deleteLogsForGroupID:(NSString *)groupID {
-  NSArray<MSLog> *logs = [self getLogsWith:groupID];
-  [self deleteLogsWith:groupID];
+  NSArray<MSLog> *logs = [self getLogsWithGroupID:groupID];
+  [self deleteLogsWithGroupID:groupID];
```
Comment (TroubleMakerBen) [inline MobileCenter/MobileCenter/Internals/Storage/MSDBStorage.m]: We are mostly using `...Id`, let's be consistent.

File: MobileCenter/MobileCenter/Internals/Storage/MSDBStorage.m [hunk 7]
```diff
@@ -45,56 +38,80 @@ - (BOOL)saveLog:(id<MSLog>)log withGroupID:(NSString *)groupID {
...
   return logs;
 }
 
-- (void)deleteLogsForId:(NSString *)logsId withGroupID:(NSString *)groupID {
+- (void)deleteLogsForId:(NSString *)batchId withGroupID:(NSString *)groupID {
```
Comment (TroubleMakerBen) [inline MobileCenter/MobileCenter/Internals/Storage/MSDBStorage.m]: It looks like `batchId` is not used here. Is this on purpose?

File: MobileCenter/MobileCenter/Internals/Storage/MSDBStorage.m [hunk 8]
```diff
@@ -45,56 +38,80 @@ - (BOOL)saveLog:(id<MSLog>)log withGroupID:(NSString *)groupID {
...
-
-  // Remove excess logs
-  if (logs.count > self.bucketFileLogCountLimit) {
-    [logs removeObjectsInRange:NSMakeRange(self.bucketFileLogCountLimit, logs.count - self.bucketFileLogCountLimit)];
+- (BOOL)loadLogsForGroupID:(NSString *)groupID
+                     limit:(NSUInteger)limit
+            withCompletion:(nullable MSLoadDataCompletionBlock)completion {
+
+  /*
+   * There is a need to determine if there will be more log available than those under the limit.
```
Comment (TroubleMakerBen) [inline MobileCenter/MobileCenter/Internals/Storage/MSDBStorage.m]: `logs` instead of `log`.

File: MobileCenter/MobileCenter/Internals/Storage/MSSqliteConnection.m [hunk 9]
```diff
@@ -1,15 +1,15 @@
+#import "MSSqliteConnection.h"
 #import <Foundation/Foundation.h>
 #import <sqlite3.h>
```
Comment (TroubleMakerBen) [inline MobileCenter/MobileCenter/Internals/Storage/MSSqliteConnection.m]: order of imports =)

File: MobileCenter/MobileCenter/Internals/Storage/MSDBStorage.m (hunk 6 above)
Comment (clpolet) [inline MobileCenter/MobileCenter/Internals/Storage/MSDBStorage.m]: Totally agree, I felt the same but didn't wanted to change GrouID to GrouId everywhere in this PR. Will address this in another PR.

File: MobileCenter/MobileCenter/Internals/Storage/MSDBStorage.m (hunk 7 above)
Comment (clpolet) [inline MobileCenter/MobileCenter/Internals/Storage/MSDBStorage.m]: Yes it's on purpose, handling this one will be another PR. I can silent the warning for now.

File: MobileCenter/MobileCenter/Internals/Model/Utils/MSUserDefaults.m (hunk 3 above)
Comment (clpolet) [inline MobileCenter/MobileCenter/Internals/Model/Utils/MSUserDefaults.m]: Good catch, didn't pay attention to this while fixing comments.

File: MobileCenter/MobileCenter/Internals/Storage/MSDBStorage.m (hunk 8 above)
Comment (clpolet) [inline MobileCenter/MobileCenter/Internals/Storage/MSDBStorage.m]: Good catch!

File: MobileCenter/MobileCenter/Internals/Storage/MSDBStorage.m [hunk 10]
```diff
@@ -31,9 +24,9 @@ - (instancetype)init {
   return self;
 }
 
--(void)initTables {
+- (void)initTables {
   NSString *createLogTableQuery = [NSString stringWithFormat:@"create table if not exists %@ (%@ text, %@ text);",
-                                   kMSLogTableName, kMSStorageKeyColumnName, kMSDataColumnName];
+                                                             kMSLogTableName, kMSGroupIDColumnName, kMSDataColumnName];
```
Comment (jaelim-ms) [inline MobileCenter/MobileCenter/Internals/Storage/MSDBStorage.m]: There is a PR which changes GroupID -> GroupId. Change it to `kMSGroupIdColumnName`

File: MobileCenter/MobileCenter/Internals/Storage/MSDBStorage.m [hunk 11]
```diff
@@ -45,56 +38,80 @@ - (BOOL)saveLog:(id<MSLog>)log withGroupID:(NSString *)groupID {
   }
   NSData *logData = [NSKeyedArchiver archivedDataWithRootObject:log];
   NSString *base64Data = [logData base64EncodedStringWithOptions:NSDataBase64EncodingEndLineWithLineFeed];
-  NSString *addLogQuery = [NSString stringWithFormat:@"insert or replace into %@ values ('%@', '%@')",
-                           kMSLogTableName, groupID, base64Data];
+  NSString *addLogQuery = [NSString
+      stringWithFormat:@"insert or replace into %@ values ('%@', '%@')", kMSLogTableName, groupID, base64Data];
```
Comment (jaelim-ms) [inline MobileCenter/MobileCenter/Internals/Storage/MSDBStorage.m]: Same here for GroupID. Please scan and rename to `groupId` for all occurrences.

File: MobileCenter/MobileCenter/Internals/Storage/MSDBStorage.m (hunk 6 above)
Comment (jaelim-ms) [inline MobileCenter/MobileCenter/Internals/Storage/MSDBStorage.m]: Well, this is a new file so it would be good to rename in this PR I guess.

File: MobileCenter/MobileCenter/Internals/Storage/MSDBStorage.m [hunk 12]
```diff
@@ -45,56 +38,80 @@ - (BOOL)saveLog:(id<MSLog>)log withGroupID:(NSString *)groupID {
...
+  [self deleteLogsWithGroupID:groupID];
   return logs;
 }
 
-- (void)deleteLogsForId:(NSString *)logsId withGroupID:(NSString *)groupID {
+- (void)deleteLogsForId:(__attribute__((unused)) NSString *)batchId withGroupID:(NSString *)groupID {
 
-  // FIXME: logsId ?
-  [self deleteLogsWith:groupID];
+  // FIXME: Restore batch deletion.
```
Comment (jaelim-ms) [inline MobileCenter/MobileCenter/Internals/Storage/MSDBStorage.m]: What does it mean? Is this still a valid FIXME?

File: MobileCenter/MobileCenter/Internals/Storage/MSDBStorage.m [hunk 13]
```diff
@@ -45,56 +38,80 @@ - (BOOL)saveLog:(id<MSLog>)log withGroupID:(NSString *)groupID {
...
+- (BOOL)loadLogsForGroupID:(NSString *)groupID
+                     limit:(NSUInteger)limit
+            withCompletion:(nullable MSLoadDataCompletionBlock)completion {
+
+  /*
+   * There is a need to determine if there will be more logs available than those under the limit.
+   * So this is just about knowing if there is at least 1 log above the limit.
+   * Thus, just 1 log is added to the requested limit and then removed later as needed.
+   */
+  NSMutableArray<MSLog> *logs = (NSMutableArray<MSLog> *)[self getLogsWithGroupID:groupID limit:(limit + 1)];
```
Comment (jaelim-ms) [inline MobileCenter/MobileCenter/Internals/Storage/MSDBStorage.m]: Do we really need this one? I think it is okay to let LogManager handles based on channel configuration.
Based on channel configuration, it will automatically pick up another batch if it has more than batch size.

File: MobileCenter/MobileCenter/Internals/Storage/MSDBStorage.m (hunk 12 above)
Comment (clpolet) [inline MobileCenter/MobileCenter/Internals/Storage/MSDBStorage.m]: It means we don't delete batches just groups. Still valid, there is a task to fix it.

File: MobileCenter/MobileCenter/Internals/Storage/MSDBStorage.m (hunk 11 above)
Comment (clpolet) [inline MobileCenter/MobileCenter/Internals/Storage/MSDBStorage.m]: Will do while merging the PR related to that change.

File: MobileCenter/MobileCenter/Internals/Storage/MSDBStorage.m (hunk 13 above)
Comment (clpolet) [inline MobileCenter/MobileCenter/Internals/Storage/MSDBStorage.m]: It's an optimization, channel will stop right there and cancel the current timer if there is no more logs from the storage to send. If we decide to change the behavior then it will be in another PR.

File: MobileCenter/MobileCenter/Internals/Storage/MSDBStorage.m (hunk 12 above)
Comment (jaelim-ms) [inline MobileCenter/MobileCenter/Internals/Storage/MSDBStorage.m]: Oh.. I see. That is a big issue. :)
